From 248e7c6417f68ad755cd2f20c484da7db2afadd6 Mon Sep 17 00:00:00 2001
From: jhenrique09 <jhenrique09.mcz@hotmail.com>
Date: Thu, 28 Dec 2017 21:42:07 -0200
Subject: [PATCH 31/41] KeyguardHostView: Auto face unlock v2

Dismiss keyguard automatically when face is recognized, updated to pie

aex edits:-
* Update our older patch
* Also revert https://github.com/AospExtended/platform_frameworks_base/commit/d055d0cb8ffbe31d3edb8ff03f7ffb5018b225da

Change-Id: If45c01af8885d57b90980551f929309101b7929d
Signed-off-by: Shubham Singh <coolsks94@gmail.com>
---
 core/java/android/provider/Settings.java        |  7 +++++++
 .../com/android/keyguard/KeyguardHostView.java  | 17 +++++++++++++++++
 .../android/keyguard/KeyguardUpdateMonitor.java |  8 ++++++++
 3 files changed, 32 insertions(+)

diff --git a/core/java/android/provider/Settings.java b/core/java/android/provider/Settings.java
index e021dea73f6..d98c2784159 100644
--- a/core/java/android/provider/Settings.java
+++ b/core/java/android/provider/Settings.java
@@ -6118,6 +6118,13 @@ public final class Settings {
          */
         public static final String KEYGUARD_SLICE_URI = "keyguard_slice_uri";
 
+        /**
+         * Enable face auto unlock on secure lock screens
+         *
+         * @hide
+         */
+        public static final String FACE_AUTO_UNLOCK = "face_auto_unlock";
+
         /**
          * Whether to speak passwords while in accessibility mode.
          *
diff --git a/packages/SystemUI/src/com/android/keyguard/KeyguardHostView.java b/packages/SystemUI/src/com/android/keyguard/KeyguardHostView.java
index aa0bcc5cf2b..375dd9e49e8 100644
--- a/packages/SystemUI/src/com/android/keyguard/KeyguardHostView.java
+++ b/packages/SystemUI/src/com/android/keyguard/KeyguardHostView.java
@@ -18,11 +18,14 @@ package com.android.keyguard;
 
 import android.app.Activity;
 import android.app.ActivityManager;
+import android.content.ContentResolver;
 import android.content.Context;
 import android.content.res.Resources;
 import android.graphics.Canvas;
 import android.media.AudioManager;
+import android.provider.Settings;
 import android.os.SystemClock;
+import android.os.UserHandle;
 import android.service.trust.TrustAgentService;
 import android.telephony.TelephonyManager;
 import android.util.AttributeSet;
@@ -63,6 +66,8 @@ public class KeyguardHostView extends FrameLayout implements SecurityCallback {
     private OnDismissAction mDismissAction;
     private Runnable mCancelAction;
 
+    private KeyguardUpdateMonitor mKeyguardUpdateMonitor;
+
     private final KeyguardUpdateMonitorCallback mUpdateCallback =
             new KeyguardUpdateMonitorCallback() {
 
@@ -95,6 +100,17 @@ public class KeyguardHostView extends FrameLayout implements SecurityCallback {
                 }
             }
         }
+
+        @Override
+        public void onTrustChanged(int userId) {
+          boolean mFaceAuto = Settings.Secure.getIntForUser(getContext().getContentResolver(),
+                           Settings.Secure.FACE_AUTO_UNLOCK, 0,
+                           UserHandle.USER_CURRENT) == 1;
+            if (userId != KeyguardUpdateMonitor.getCurrentUser()) return;
+            if (mKeyguardUpdateMonitor.getUserCanSkipBouncer(userId) && mKeyguardUpdateMonitor.getUserHasTrust(userId) && mKeyguardUpdateMonitor.isFaceTrusted() && mFaceAuto) {
+                dismiss(false, userId);
+            }
+        }
     };
 
     // Whether the volume keys should be handled by keyguard. If true, then
@@ -112,6 +128,7 @@ public class KeyguardHostView extends FrameLayout implements SecurityCallback {
 
     public KeyguardHostView(Context context, AttributeSet attrs) {
         super(context, attrs);
+        mKeyguardUpdateMonitor = KeyguardUpdateMonitor.getInstance(context);
         KeyguardUpdateMonitor.getInstance(context).registerCallback(mUpdateCallback);
     }
 
diff --git a/packages/SystemUI/src/com/android/keyguard/KeyguardUpdateMonitor.java b/packages/SystemUI/src/com/android/keyguard/KeyguardUpdateMonitor.java
index 04206b9d4a0..8596f2eab1b 100644
--- a/packages/SystemUI/src/com/android/keyguard/KeyguardUpdateMonitor.java
+++ b/packages/SystemUI/src/com/android/keyguard/KeyguardUpdateMonitor.java
@@ -252,6 +252,9 @@ public class KeyguardUpdateMonitor implements TrustManager.TrustListener,
     private static final int HW_UNAVAILABLE_TIMEOUT = 3000; // ms
     private static final int HW_UNAVAILABLE_RETRY_MAX = 3;
 
+    // For face unlock identification
+    private String lastBroadcastActionReceived;
+
     private final Handler mHandler = new Handler(Looper.getMainLooper()) {
         @Override
         public void handleMessage(Message msg) {
@@ -854,11 +857,16 @@ public class KeyguardUpdateMonitor implements TrustManager.TrustListener,
         }
     };
 
+    public boolean isFaceTrusted(){
+        return lastBroadcastActionReceived.equals(ACTION_FACE_UNLOCK_STOPPED);
+    }
+
     private final BroadcastReceiver mBroadcastAllReceiver = new BroadcastReceiver() {
 
         @Override
         public void onReceive(Context context, Intent intent) {
             final String action = intent.getAction();
+            lastBroadcastActionReceived = action;
             if (AlarmManager.ACTION_NEXT_ALARM_CLOCK_CHANGED.equals(action)) {
                 mHandler.sendEmptyMessage(MSG_TIME_UPDATE);
             } else if (Intent.ACTION_USER_INFO_CHANGED.equals(action)) {
-- 
2.17.1

